codeworld-compiler/src/ParseCode.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+{-
+  Copyright 2017 The CodeWorld Authors. All rights reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-}
+
+module ParseCode (checkParsedCode) where
+
+import qualified Data.ByteString as B
+import           Data.ByteString.Char8 (pack)
+import           Data.List (intercalate)
+import           Data.List.Split (splitOn)
+import           Data.Generics
+import           Language.Haskell.Exts
+
+checkParsedCode :: FilePath -> FilePath -> IO Bool
+checkParsedCode src err = do
+    result <- parseFile src
+    case result of
+        ParseOk mod -> case getErrors mod of
+            [] -> return True
+            errors -> writeFile err (intercalate "\n\n" errors) >> return False
+        ParseFailed _ _ -> return True
+
+getErrors :: Module SrcSpanInfo -> [String]
+getErrors m = everything (++) (mkQ [] badExps) m ++
+              everything (++) (mkQ [] badMatches) m
+
+badExps :: Exp SrcSpanInfo -> [String]
+badExps (App x (Var l _) e) | not (isGoodExpRhs e) = [
```

Comment:
You can't limit to a Var on the left side of an application.  At the very least, you need to match Con, Lambda, LeftSection, RightSection... but why not just match anything, and use the location from the App?

Reply:
If we take the starting point from the location of the App, then it will show, missing parentheses at the starting of the function name rather than at the end of it, which might confuse the user. One thing we can do is, keep this and add one more line like this:
```
badExps (App x _ e) | not (isGoodExpRhs e) = [
```
for other cases.

What say?

Reply:
Actually, it will do that anyway, since formatLocation uses the start of the span.  Anyway, it's showing the starting point of the function application that is missing the parentheses, which seems logical to me.

Maybe there's some tweaking to be done, but I would rather be correct first, and tweak later, rather than start out with a bunch of kludges.